packages/lib/src/components/ThreeDS2/components/DeviceFingerprint/PrepareFingerprint3DS2.tsx

```diff
@@ -55,18 +64,90 @@ class PrepareFingerprint3DS2 extends Component<PrepareFingerprint3DS2Props, Prep
+                // if (this.props.isMDFlow) {}
+
+                console.debug('### PrepareFingerprint3DS2::exiting:: no valid threeDSMethodURL');
+                return;
+            }
+
+            const hasValid3DSMethodNotificationURL = isValidHttpUrl(threeDSMethodNotificationURL, shouldAllowHttpDomains);
+            const hasValidPostMessageDomain = isValidHttpUrl(postMessageDomain, shouldAllowHttpDomains);
+            const hasTransServerID = threeDSServerTransID?.length;
+
+            if (!hasValid3DSMethodNotificationURL || !hasValidPostMessageDomain || !hasTransServerID) {
+                /**
+                 * NOTE: EMVCo_3DS_CoreSpec_v2.3.1_20220831.pdf states that we should return: threeDSCompInd: 'N'
+                 * when the fingerprinting process "Did not run or did not successfully complete"
+                 */
+                this.setStatusComplete(
+                    { threeDSCompInd: 'N' },
+                    {
+                        errorCode: Analytics3DS2Errors.TOKEN_IS_MISSING_OTHER_PROPS,
+                        message: `${THREEDS2_FINGERPRINT_ERROR}: Decoded token is missing one or more of the following properties (threeDSMethodNotificationURL | postMessageDomain | threeDSServerTransID)`
+                    }
+                );
+
+                console.debug(
+                    '### PrepareFingerprint3DS2::exiting:: Decoded token is missing one or more of the following properties (threeDSMethodNotificationURL | postMessageDomain | threeDSServerTransID)'
+                );
+
+                return;
+            }
+
+            // Proceed to allow component to render
+            this.setState({ status: 'retrievingFingerPrint' });
+            //
+        } else {
+            // Only render component if we have fingerPrintData. Otherwise, complete with threeDSCompInd: 'N'
+
+            const errorMsg: string = (this.state.fingerPrintData as ErrorObject).error;
+
+            const errorCode =
+                errorMsg.indexOf(MISSING_TOKEN_IN_ACTION_MSG) > -1
+                    ? Analytics3DS2Errors.ACTION_IS_MISSING_TOKEN
+                    : Analytics3DS2Errors.TOKEN_DECODE_OR_PARSING_FAILED;
 
-        // Render
-        this.setState({ status: 'retrievingFingerPrint' });
+            this.setStatusComplete(
+                { threeDSCompInd: 'N' },
+                {
+                    errorCode,
+                    message: `${THREEDS2_FINGERPRINT_ERROR}: ${errorMsg}` // can be: 'Missing "token" property from threeDS2 action', 'not base64', 'malformed URI sequence' or 'Could not JSON parse token'
+                }
+            );
+
+            console.debug('### PrepareFingerprint3DS2:: token does not exist or could not be base64 decoded &/or JSON.parsed');
+        }
     }
 
-    setStatusComplete(resultObj: ResultObject) {
+    setStatusComplete(resultObj: ResultObject, errorCodeObject: ErrorCodeObject = null) {
         this.setState({ status: 'complete' }, () => {
```

Comment:
Just out of curiosity, why writing it inside the `setState` callback?

Reply:
Good question. Don't remember -it's just always been done this way. I presume there's some subtle difference in timing - doing it this way means you change state, _then_ run the callback, and _then_ do the re-render (triggered by the fact that state has changed)?